File: opspilot-enterprise/packages/shared-schema/src/opspilot_schema/evidence.py
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel


EvidenceSourceType = Literal["event", "metric", "log", "topology", "kb", "change", "external_kb", "detail", "alert"]


class Evidence(BaseModel):
    evidence_id: str
    source: str
    source_type: EvidenceSourceType
    object_type: str
    object_id: str
    timestamp: str
    summary: str
    raw_ref: str | None = None
    confidence: float
    correlation_key: str | None = None


class EvidenceSourceStats(BaseModel):
    source_type: EvidenceSourceType
    count: int
    avg_confidence: float


class EvidenceCoverage(BaseModel):
    requested_sources: int
    collected_sources: int
    missing_sources: int
    required_evidence_types: list[str] = []
    present_evidence_types: list[str] = []
    missing_critical_evidence: list[str] = []
    sufficiency_score: float = 0.0
    freshness_score: float = 0.0


class EvidenceError(BaseModel):
    source: str
    message: str


class EvidenceContradiction(BaseModel):
    type: str
    summary: str
    evidence_refs: list[str] = []
    severity: Literal["high", "medium", "low"] = "medium"


class EvidencePackage(BaseModel):
    package_id: str
    incident_id: str | None = None
    session_id: str | None = None
    evidences: list[Evidence] = []
    created_at: str
    source_stats: list[EvidenceSourceStats] = []
    coverage: EvidenceCoverage | None = None
    errors: list[EvidenceError] = []
    required_evidence_types: list[str] = []
    present_evidence_types: list[str] = []
    missing_critical_evidence: list[str] = []
    sufficiency_score: float = 0.0
    freshness_score: float = 0.0
    contradictions: list[EvidenceContradiction] = []

File: opspilot-enterprise/packages/shared-types/src/evidence.ts
export interface Evidence {
  evidence_id: string;
  source: string;
  source_type: "event" | "metric" | "log" | "topology" | "kb" | "change" | "external_kb" | "detail" | "alert";
  object_type: string;
  object_id: string;
  timestamp: string;
  summary: string;
  raw_ref: string | null;
  confidence: number;
  correlation_key: string | null;
}

export interface EvidenceSourceStats {
  source_type: Evidence["source_type"];
  count: number;
  avg_confidence: number;
}

export interface EvidenceCoverage {
  requested_sources: number;
  collected_sources: number;
  missing_sources: number;
  required_evidence_types: string[];
  present_evidence_types: string[];
  missing_critical_evidence: string[];
  sufficiency_score: number;
  freshness_score: number;
}

export interface EvidenceError {
  source: string;
  message: string;
}

export interface EvidenceContradiction {
  type: string;
  summary: string;
  evidence_refs: string[];
  severity: "high" | "medium" | "low";
}

export interface EvidencePackage {
  package_id: string;
  incident_id: string | null;
  session_id: string | null;
  evidences: Evidence[];
  created_at: string;
  source_stats?: EvidenceSourceStats[];
  coverage?: EvidenceCoverage;
  errors?: EvidenceError[];
  required_evidence_types?: string[];
  present_evidence_types?: string[];
  missing_critical_evidence?: string[];
  sufficiency_score?: number;
  freshness_score?: number;
  contradictions?: EvidenceContradiction[];
}
